crates/algorithms/src/haversine/mod.rs

```diff
@@ -0,0 +1,131 @@
+                #[doc = "* `p2`: A [`Point2`] representing the second geographical point."]
+                #[doc = "* `sphere_radius`: The radius of the sphere, typically the Earth's radius in kilometers or miles."]
+                #[doc = ""]
+                #[doc = "# Returns"]
+                #[doc = "A 'float', the distance between `p1` and `p2` along the surface of the sphere, using the Haversine formula."]
+                pub fn [<calculate_haversine_distance_ $prec>](p1: nalgebra::Point2<$prec>, p2: nalgebra::Point2<$prec>, sphere_radius: $prec) -> $prec {
+
+                    calculate_haversine_distance(p1,p2,sphere_radius)
+                }
+                #[doc = "Calculates the initial bearing (forward azimuth) from the first point to the second point."]
+                #[doc = ""]
+                #[doc = "This function computes the initial bearing, or forward azimuth, between two points on the surface"]
+                #[doc = "of a sphere, assuming a spherical Earth model. The bearing is the direction one must travel"]
+                #[doc = "from the first point to reach the second point, expressed as an angle from North (0 degrees)"]
+                #[doc = "in a clockwise direction."]
+                #[doc = ""]
+                #[doc = "# Arguments"]
+                #[doc = "* `p1`: A [`Point2`] representing the starting geographical point (latitude and longitude)."]
+                #[doc = "* `p2`: A [`Point2`] representing the destination geographical point (latitude and longitude)."]
+                #[doc = ""]
+                #[doc = "# Type Parameters"]
+                #[doc = "* `T`: The floating point type used for the calculation, which must implement `Copy`, `Float`, and `Debug`."]
+                #[doc = "  This allows for a generic implementation that can work with different numeric types, such as `f32` or `f64`."]
+                #[doc = ""]
+                #[doc = "# Returns"]
+                #[doc = "* A value that representing the initial bearing from `p1` to `p2`, in degrees. The result is normalized"]
+                #[doc = "  to a range of 0 to 360 degrees."]
+                pub fn [<return_bearing_between_points_ $prec>](p1: nalgebra::Point2<$prec>, p2: nalgebra::Point2<$prec>) -> $prec {
+
+                    return_bearing_between_points(p1,p2)}
+                }
+
+
+            }
+        };}
+#[cfg(feature = "pregenerated")]
+impl_p_i_p_algorithm!(f32, single);
+#[cfg(feature = "pregenerated")]
+impl_p_i_p_algorithm!(f64, double);
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_distance() {
+        let p1 = Point2::new(52.5200, 13.4050); // Berlin, Germany
+        let p2 = Point2::new(48.8566, 2.3522); // Paris, France
+        let earth_radius_km = 6371.0;
+        let distance = calculate_haversine_distance(p1, p2, earth_radius_km);
+        let expected_distance = 878.0; // Approximate distance in km
```

Comment:
Tested with various harvesine calculators, when accounting for earth radius as 6371.0, the distance should be 877.46km,
I think the assert should have a much lower threshold, maybe 0.01